src/shared/vscode/commands2.ts

```diff
@@ -367,42 +368,143 @@ interface CommandInfo<T extends Callback> {
     readonly telemetryThrottleMs?: number | false
 }
 
-// This debouncing is a temporary safe-guard to mitigate the risk of instrumenting all
-// commands with telemetry at once. The alternative is to do a gradual rollout although
-// it's not much safer in terms of weeding out problematic commands.
-const telemetryInfo = new Map<string, { startTime: number; debounceCount: number }>()
-
-function getInstrumenter(id: string, threshold: number, telemetryName?: MetricName) {
-    const currentTime = Date.now()
-    const info = telemetryInfo.get(id)
+function getInstrumenter(id: { id: string; args: any[] }, threshold: number, telemetryName?: MetricName) {
+    const currentTime = globals.clock.Date.now()
+    const info = TelemetryDebounceInfo.instance.get(id)
 
+    // to reduce # of events actually emitted, we throttle the same* event for brief period of time
+    // and instead increment a counter for executions while throttled.
     if (!telemetryName && info?.startTime !== undefined && currentTime - info.startTime < threshold) {
         info.debounceCount += 1
-        telemetryInfo.set(id, info)
-        getLogger().debug(`commands: skipped telemetry for "${id}"`)
+        TelemetryDebounceInfo.instance.set(id, info)
+        getLogger().debug(`commands: skipped telemetry for "${id.id}"`)
 
         return undefined
     }
 
     // Throttling occurs regardless of whether or not the instrumenter is invoked
     const span = telemetryName ? telemetry[telemetryName] : telemetry.vscode_executeCommand
     const debounceCount = info?.debounceCount !== 0 ? info?.debounceCount : undefined
-    telemetryInfo.set(id, { startTime: currentTime, debounceCount: 0 })
+    TelemetryDebounceInfo.instance.set(id, { startTime: currentTime, debounceCount: 0 })
 
     return <T extends Callback>(fn: T, ...args: Parameters<T>) =>
         span.run(span => {
-            ;(span as Metric<VscodeExecuteCommand>).record({ command: id, debounceCount })
+            ;(span as Metric<VscodeExecuteCommand>).record({
+                command: id.id,
+                debounceCount,
+                source: BaseCommandSource.findSource(args),
+            })
 
             return fn(...args)
         })
 }
 
+/**
+ * Adding an instance of this to the execution args of your {@link RegisteredCommand}
+ * will set the "source" attribute in the automatically emitted `vscode_executeCommand` metric.
+ */
+export class BaseCommandSource {
+    constructor(readonly source: string) {}
+
+    /**
+     * Returns the `source` of a {@link BaseCommandSource} if found
+     * in the given args.
+     */
+    static findSource(args: any[]): string | undefined {
+        const source = args.find(arg => arg instanceof BaseCommandSource)
```

Comment:
This is technically possible, but unlikely since it doesn't make sense in this context to ask for more than one.

> Is there no predictable arg name that we could use instead?
Not sure what you mean by this. Can we determine by a name instead of `instanceof`?

Reply:
Yes, if we could just always look for `args.source` or something.